specifyweb/frontend/js_src/lib/components/TreeView/AddRank.tsx

```diff
@@ -0,0 +1,158 @@
+import { interactionsText } from '../../localization/interactions';
+import { Select } from '../Atoms/Form';
+import { RA } from '../../utils/types';
+import { ResourceView } from '../Forms/ResourceView';
+import { SpecifyResource } from '../DataModel/legacyTypes';
+import { GeographyTreeDefItem } from '../DataModel/types';
+import { strictIdFromUrl } from '../DataModel/resource';
+
+export function AddRank<SCHEMA extends AnyTree>({
+  tableName,
+  treeDefinitionItems,
+}: {
+  readonly tableName: SCHEMA['tableName'];
+  readonly treeDefinitionItems: RA<
+    SerializedResource<FilterTablesByEndsWith<'TreeDefItem'>>
+  >;
+}): JSX.Element {
+  const loading = React.useContext(LoadingContext);
+
+  const [isAddingRank, setIsAddingRank] = React.useState(false);
+  const [isAddingParentRank, setIsAddingParentRank] = React.useState(false);
+
+  const [parentRank, setParentRank] = React.useState('');
+  const treeId = strictIdFromUrl(treeDefinitionItems[0].treeDef);
+
+  const treeResource = React.useMemo(
+    () =>
+      tableName === 'Geography'
+        ? new tables.GeographyTreeDefItem.Resource({
+            _tableName: 'GeographyTreeDefItem',
+          })
+        : tableName === 'Taxon'
+        ? new tables.TaxonTreeDefItem.Resource({
+            _tableName: 'TaxonTreeDefItem',
+          })
+        : tableName === 'LithoStrat'
+        ? new tables.LithoStratTreeDefItem.Resource({
+            _tableName: 'LithoStratTreeDefItem',
+          })
+        : new tables.StorageTreeDefItem.Resource({
+            _tableName: 'StorageTreeDefItem',
+          }),
+    [tableName]
+  );
+
+  function addRank(): void {
+    const url = `/api/specify_tree/${tableName.toLowerCase()}/add_tree_rank/`;
+    loading(
+      ping(url, {
+        method: 'POST',
+        body: {
+          newRankName: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('name'),
+          parentRankName: parentRank,
+          treeID: treeId,
+          newRankTitle: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('title'),
+          remarks: (treeResource as SpecifyResource<GeographyTreeDefItem>).get(
+            'remarks'
+          ),
+          textAfter: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('textAfter'),
+          textBefore: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('textBefore'),
+          isEnforced: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('isEnforced'),
+          isInFullName: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('isInFullName'),
+          fullNameSeparator: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('fullNameSeparator'),
+        },
+      }).then(() => globalThis.location.reload())
+    );
+    setIsAddingRank(false);
+  }
```

Comment:
none of this should be necessary - you should just let ResourceView save the resource like it would normally do, and the back-end business rules should take care of the rest
if there is a back-end business rule violation, back-end will throw, but the front-end should be able to handle that gracefully already thanks to the changes Jason did